File: solutions/python/knapsack/1/knapsack.py
def maximum_value(maximum_weight, items):
    n = len(items)
    dp = [[0 for _ in range(maximum_weight + 1)] for _ in range(n + 1)]
    for i in range(1,n+1,1):
        weight = items[i - 1]["weight"]
        value = items[i - 1]["value"]
        for w in range(maximum_weight + 1):
            if weight <= w:
                dp[i][w] = max(dp[i - 1][w], dp[i-1][w - weight] + value)
            else:
                dp[i][w] = dp[i - 1][w]

    return dp[n][maximum_weight]

File: solutions/cpp/knapsack/1/knapsack.cpp
#include "knapsack.h"

namespace knapsack {
    int max(const int a, const int b) {
        return a > b ? a : b;
    }

    int maximum_value(const int max_weight, const std::vector<Item> &items) {
        if (items.empty() || max_weight == 0) {
            return 0;
        }
        std::vector<std::vector<int> > dp(items.size() + 1, std::vector<int>(max_weight + 1, 0));

        for (int i = 1; i <= static_cast<int>(items.size()); i++) {
            const int weight = items[i - 1].weight;
            const int value = items[i - 1].value;
            for (int w = 1; w <= max_weight; w++) {
                if (weight <= w) {
                    const int take = dp[i - 1][w -weight] + value;
                    const int skip = dp[i - 1][w];
                    dp[i][w] = max(take, skip);
                }else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }
        return dp[items.size()][max_weight];
    }
} // namespace knapsack
